lib/vnet/ssh_proxy.go

```diff
@@ -0,0 +1,254 @@
+// between 2 established SSH connections. serverConn represents an incoming SSH
+// connection where this proxy acts as a server, client represents an outgoing
+// SSH connection where this proxy acts as a client.
+func proxySSHConnection(
+	ctx context.Context,
+	serverConn sshConn,
+	clientConn sshConn,
+) {
+	closeConnections := sync.OnceFunc(func() {
+		clientConn.conn.Close()
+		serverConn.conn.Close()
+	})
+	// Close both connections if the context is canceled.
+	context.AfterFunc(ctx, closeConnections)
+
+	// Avoid leaking goroutines by tracking them with a waitgroup.
+	// If any task exits make sure to close both connections so that all other
+	// tasks can terminate.
+	var wg sync.WaitGroup
+	runTask := func(task func()) {
+		wg.Add(1)
+		go func() {
+			task()
+			closeConnections()
+			wg.Done()
+		}()
+	}
+
+	// Proxy channels initiated by either connection.
+	runTask(func() {
+		proxyChannels(ctx, serverConn.conn, clientConn.chans, closeConnections)
+	})
+	runTask(func() {
+		proxyChannels(ctx, clientConn.conn, serverConn.chans, closeConnections)
+	})
+
+	// Proxy global requests in both directions.
+	runTask(func() {
+		proxyGlobalRequests(ctx, serverConn.conn, clientConn.reqs, closeConnections)
+	})
+	runTask(func() {
+		proxyGlobalRequests(ctx, clientConn.conn, serverConn.reqs, closeConnections)
+	})
+
+	wg.Wait()
+}
+
+func proxyChannels(
+	ctx context.Context,
+	targetConn ssh.Conn,
+	chans <-chan ssh.NewChannel,
+	closeConnections func(),
+) {
+	// Proxy each SSH channel in its own goroutine, make sure they don't leak by
+	// tracking with a WaitGroup.
+	var wg sync.WaitGroup
+	for newChan := range chans {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			proxyChannel(ctx, targetConn, newChan, closeConnections)
+		}()
+	}
+	wg.Wait()
+}
+
+func proxyChannel(
+	ctx context.Context,
+	targetConn ssh.Conn,
+	newChan ssh.NewChannel,
+	closeConnections func(),
+) {
+	log := log.With("channel_type", newChan.ChannelType())
+	log.DebugContext(ctx, "Proxying new SSH channel")
+
+	// Try to open a corresponding channel on the target.
+	targetChan, targetChanRequests, err := targetConn.OpenChannel(
+		newChan.ChannelType(), newChan.ExtraData())
+	if err != nil {
+		var openChannelErr *ssh.OpenChannelError
+		var rejectErr error
+		if errors.As(err, &openChannelErr) {
+			// The target rejected the channel, this is totally expected in some
+			// cases, just reject the incoming channel request.
+			rejectErr = trace.Wrap(newChan.Reject(openChannelErr.Reason, openChannelErr.Message))
+		} else {
+			// We got an unexpected error trying to open the channel on the
+			// target, this is fatal, log and kill the connection.
+			log.DebugContext(ctx, "Unexpected error opening SSH channel on target",
+				"error", err)
+			msg := "unexpected error opening channel on target: " + err.Error()
+			rejectErr = trace.Wrap(newChan.Reject(ssh.ConnectionFailed, msg))
+			closeConnections()
```

Comment:
I would argue that this is incorrect, we don't know if the channel would've gotten rejected, the client might make decisions based on the rejection that are outside of the immediate use of the SSH connection, and `ConnectionFailed` might not be at all appropriate for the channel.

Since it's not like closing the connection immediately after rejecting the channel changes much:
```suggestion
			closeConnections()
			// "resolve" the NewChannel even after closing the connections
			_ = newChan.Reject(0, "")
			return
```

Reply:
sure, in the latest it first closes the connections then rejects then calls newChan.Reject with zero values